src/ai/backend/storage/context.py

```diff
@@ -170,28 +172,37 @@ def list_volumes(self) -> Mapping[str, VolumeInfo]:
         return {name: VolumeInfo(**info) for name, info in self.local_config["volume"].items()}
 
     async def __aexit__(self, *exc_info) -> Optional[bool]:
+        for volume in self.volumes.values():
+            await volume.shutdown()
+
         await self.storage_plugin_ctx.cleanup()
         await self.manager_webapp_plugin_ctx.cleanup()
         await self.client_webapp_plugin_ctx.cleanup()
         return None
 
     @actxmgr
     async def get_volume(self, name: str) -> AsyncIterator[AbstractVolume]:
-        try:
-            volume_config = self.local_config["volume"][name]
-        except KeyError:
-            raise InvalidVolumeError(name)
-        volume_cls: Type[AbstractVolume] = self.backends[volume_config["backend"]]
-        volume_obj = volume_cls(
-            local_config=self.local_config,
-            mount_path=Path(volume_config["path"]),
-            options=volume_config["options"] or {},
-            etcd=self.etcd,
-            event_dispathcer=self.event_dispatcher,
-            event_producer=self.event_producer,
-        )
-        await volume_obj.init()
-        try:
-            yield volume_obj
-        finally:
-            await volume_obj.shutdown()
+        if name in self.volumes:
+            yield self.volumes[name]
+        else:
+            try:
+                volume_config = self.local_config["volume"][name]
+            except KeyError:
+                raise InvalidVolumeError(name)
+            volume_cls: Type[AbstractVolume] = self.backends[volume_config["backend"]]
+            volume_obj = volume_cls(
+                local_config=self.local_config,
+                mount_path=Path(volume_config["path"]),
+                options=volume_config["options"] or {},
+                etcd=self.etcd,
+                event_dispathcer=self.event_dispatcher,
+                event_producer=self.event_producer,
+            )
+
+            await volume_obj.init()
+            self.volumes[name] = volume_obj
```

Comment:
Does it work with high availability setups?

Reply:
No, in the HA setup, each cache will be stored separately on the storage proxy servers.
Do all storage proxy servers need to be implemented to use the same storage as cache?

Reply:
In terms of memoization I don't think this objects do not have to (and also cannot be) shared among every nodes consisting HA setup.